### PR TITLE
Automatically scale the number of workers as workload comes and goes

### LINKED ISSUE #1581: Scale down pipelines rcs when pipeline is not being run
This gets more important once #1553 is in because the pipeline's pods can block other things from scheduling.

### PR DESCRIPTION
**Pull request by derekchiang**
Fix #1581

### CODE REVIEW

File: doc/reference/pipeline_spec.md [hunk 1]
```diff
@@ -191,6 +192,14 @@ store such as s3, Google Cloud Storage or Azure Storage. Data will be pushed
 after the user code has finished running but before the job is marked as
 successful.
 
+## Scale-down threshold (optional)
+
+`scaleDownThreshold` specifies when the worker pods of a pipeline should be terminated.
+
+By default, a pipeline’s worker pods are always running.  When `scaleDownThreshold` is set, the worker pods are terminated after they have been idle for the given duration.  When a new input commit comes in, the worker pods are then re-created.
+
+`scaleDownThreshold` is a string that needs to be sequence of decimal numbers, each with optional fraction and a unit suffix, such as “300ms”, “1.5h” or “2h45m”. Valid time units are “s”, “m”, “h”.
```
Comment (JoeyZwicker) [inline doc/reference/pipeline_spec.md]: If fractions are actually valid, can you add an example?

**Comment by JoeyZwicker** [comment]
One comment, but docs side of things LGTM. I'll need to add this to our production section

File: src/server/pps/server/commit_set.go [hunk 2]
```diff
@@ -79,19 +60,72 @@ func newBranchSetFactory(ctx context.Context, pfsClient pfs.APIClient, inputs []
...
 						Name: branchName,
 						Head: commitInfo.Commit,
+					}:
 					}
 				}
 			}(branchName)
 		}
 	}
 
+	ch := make(chan *branchSet)
```
Comment (msteffen) [inline src/server/pps/server/commit_set.go]: Just to make sure I understand this code: is this part of the scaling change, or just more general refactoring?

File: src/server/cmd/worker/main.go [hunk 3]
```diff
@@ -169,8 +169,8 @@ func do(appEnvObj interface{}) error {
 
 	// Prepare to write "key" into etcd by creating lease -- if worker dies, our
 	// IP will be removed from etcd
-	ctx, _ := context.WithTimeout(context.Background(), 30*time.Second)
-	resp, err := etcdClient.Grant(ctx, 60 /* seconds */)
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
+	resp, err := etcdClient.Grant(ctx, 10 /* seconds */)
```
Comment (derekchiang) [inline src/server/cmd/worker/main.go]: Yeah that's basically it.  Technically having the old IPs around shouldn't stop new jobs from completing.  We'd just have a couple zombie worker goros on the server side that get shut down once the old IPs expire.

File: src/server/pps/server/api_server.go [hunk 4]
```diff
@@ -986,6 +986,90 @@ func isNotFoundErr(err error) bool {
...
+			break
+		}
+		if jobInfo.PipelineID == pipelineInfo.ID {
+			jobInfos = append(jobInfos, &jobInfo)
+		}
+	}
+	return jobInfos, nil
+}
+
+// signalJobCompletion waits for a job to complete and then sends the job back on jobCompletionCh.
```
Comment (derekchiang) [inline src/server/pps/server/api_server.go]: Yeah that makes sense.  Will do.

File: src/server/pps/server/commit_set.go (hunk 2 above)
Comment (derekchiang) [inline src/server/pps/server/commit_set.go]: It was necessary to refactor it such that `branchSetFactory` returns a channel so `jobManager` can use `select` to block on either new input commits or job completion signals.

**Pull request merged**